backend/internal/post/service.go

```diff
@@ -0,0 +1,86 @@
+package post
+
+import (
+	"backend/internal"
+	"backend/internal/database"
+	"context"
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v5/pgtype"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+)
+
+//go:generate mockery --name Querier
+type Querier interface {
+	FindAll(ctx context.Context) ([]Post, error)
+	FindByID(ctx context.Context, id uuid.UUID) (Post, error)
+	Create(ctx context.Context, arg CreateParams) (Post, error)
+	Delete(ctx context.Context, id uuid.UUID) error
+	Update(ctx context.Context, arg UpdateParams) (Post, error)
+}
+
+type Service struct {
+	logger *zap.Logger
+	tracer trace.Tracer
+	query  Querier
+}
+
+func NewService(logger *zap.Logger, db *pgxpool.Pool) Service {
+	return Service{
+		logger: logger,
+		tracer: otel.Tracer("post/service"),
+		query:  New(db),
+	}
+}
+
+func (s Service) GetAll(ctx context.Context) ([]Post, error) {
+	traceCtx, span := s.tracer.Start(ctx, "GetByID")
+	defer span.End()
+	logger := internal.LoggerWithContext(traceCtx, s.logger)
+
+	posts, err := s.query.FindAll(traceCtx)
+	if err != nil {
+		err = database.WrapDBError(err, logger)
+		span.RecordError(err)
+		logger.Error("Failed to get all posts", zap.Error(err))
```

Comment:
We should aim to keep the `if err != nil` block lean and avoid adding too much logic there. Currently, `WrapDBError` already takes a `logger` and logs the error details, so the additional `logger.Error("Failed to get all posts", zap.Error(err))` feels redundant.

Instead of duplicating logging in the caller, we could enhance `WrapDBError` to accept an operation description (e.g., "Failed to get all posts") and log it there alongside the error. This would centralize error handling, reduce repetition, and keep the calling code cleaner.

I know you may have followed my previous style; I’ll be adjusting my code to match this pattern too.